draft-pauly-taps-arch.md

```diff
@@ -205,103 +205,93 @@ Beyond the basic objects, there are several high-level groups of actions that an
+ +-->| Preconnection |----------->| Connection |---------------> Closed
+ |   +---------------+     :      +------------+ Connection:
+ |                         :      ^   ^    |     Finished  :
+ +-- Local Endpoint        :      |   |    |               :
+ |                         :      |   |    +-----+         :
+ +-- Remote Endpoint       :      |   |          |Content  :
+ |                         :      |   |Send()    |Received :
+ +-- Path Selection        :      | +---------+  |         :
+ |   Properties            :      | | Content |<-+         :
+ |                         :      | +---------+            :
+ +-- Protocol Selection    :      |                        :
+ |   Properties            :      |                        :
+ |                         :      |                        :
+ +-- Specific Protocol     :      |                        :
+ |   Properties            :      |                        :
+ |                         :      |                        :
+ |   +----------+          :      |                        :
+ +-->| Listener |-----------------+                        :
+     +----------+ Connection Received                      :
+           ^               :                               :
+           |               :                               :
+        Listen()           :                               :
+~~~~~~~~~~
+{: #fig-lifetime title="The lifetime of a connection"}
 
-Connection
+### Basic Objects {#objects}
 
->> A Connection is the fundamental object used by an application for all interaction with a peer and data transfer. It is generally capable of bi-directional communication. It has state that represents the capability of being able to send or receive content between its local and remote endpoints. This capability may be based merely on the existence of a path between the two endpoints and an application's permission to send and receive on its local endpoint; or may represent successful protocol handshakes at one or more layers that are pre-requisites to receiving content.
+* Preconnection: A Preconnection object is a representation of a potential connection. It has state that describes parameters of a Connection that might exist in the future: the Local Endpoint from which that Connection will be established, the Remote Endpoint to which it will connect, and Path Selection Properties, Protocol Selection Properties, and Specific Protocol Properties that influence the choice of transport that Connection will use. A Preconnection can be fully specified and represent a single possible Connection, or it can be partially specified such that it represents a family of possible Connections. 
 
-Listener
+* Connection: A Connection object represents an active transport protocol instance that can send and/or receive Content between a Local Endpoint and a Remote Endpoint. It holds state pertaining to the underlying transport protocol instance and any ongoing data transfer. 
```

Comment:
I'd like to not lose the original intent of the text before, that captured the idea that this "Connection" still works for stateless protocols like UDP, just by maintaining the five-tuple path. A note to say that the readiness of the connection to send and receive may or may not be based on protocol handshakes would be good.

Reply:
Also i guess e.g. DNS resolution in done by the connection (not the preconception), right? In this case there a change that the connection fails even before an active transport protocol instance has been established. And the other question is, can a connection have more than one active protocol instance? I'd say yes, as you could emulate multipath by using e.g. two TCP connections.

Reply:
Replying to myself.... I guess DNS resolution is done during Initiate() and if that fails no connection is established, right? I still wondering if there are cases where you have a connection but no active protocol instance (yet or anymore)...

Reply:
About the "you could emulate multipath by using e.g. two TCP connections" bit: this is exactly the email conversation I just have / had with Tommy - I think that should be possible, but it would need us to add something to the API such that the application can say "my peer can handle this".

Reply:
@tfpauly good point, will fix
@mirjak DNS resolution is done in Initiate(), and you'll get an error event if it doesn't resolve to something you can connect to. I haven't thought whether a Connection can support multiple underlying protocol instances, but the case where there is no longer an active protocol instance seems needed to represent a closed connection.

Reply:
re resolution: I think whether this happens before or after Initiate() should be left up to an implementation. I could certainly see implementations that take action during Preconnect() in order to reduce Initiate() latency, and warming up the local name cache is one of those possible actions...

Reply:
@britram this may be an optimisation hint we allow, but I'd like when resolution occurs to be deterministic by default